### PR TITLE
solves issue #706

### PR DESCRIPTION
**Pull request by deepanshu44**
create-neon now deletes the partial project directory on a non-zero exit 
#706

### CODE REVIEW

File: create-neon/src/shell.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
-import { spawn } from 'child_process';
-
+import { spawn } from "child_process";
+import fs from "fs/promises";
```
Comment (kjvalencik) [inline create-neon/src/shell.ts:2]: Unfortunately, the promises API is only exposed at `fs/promises` in newer node versions. In order to support older versions, this will need to be updated:

```suggestion
import { promises as fs } from "fs";
```

This can be seen in the failed CI unit tests.

File: create-neon/src/bin/create-neon.ts [hunk 2]
```diff
@@ -30,7 +31,12 @@ function inferVersions(): Versions {
 }
 
 async function main(name: string) {
+  let workingDirPath: string = process.cwd()
+  let tmpDirPath: string = tmpdir()
+
   try {
+    //change PWD to temp
+    process.chdir(tmpDirPath)
     await fs.mkdir(name);
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:40]: In order to avoid potential collisions in the `tmpdir` can you use a random directory name?

https://nodejs.org/api/fs.html#fspromisesmkdtempprefix-options

File: create-neon/src/bin/create-neon.ts [hunk 3]
```diff
@@ -40,7 +46,9 @@ async function main(name: string) {
 
   try {
     pkg = await Package.create(name);
+
   } catch (err) {
+    await fs.rm(process.argv[2], { recursive: true, force: true })
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:51]: This should use `name` instead of the `argv` directly.

Also, I think it would be better to move the `fs.rm` into `die`.

File: create-neon/src/bin/create-neon.ts [hunk 4]
```diff
@@ -53,7 +61,14 @@ async function main(name: string) {
       versions: inferVersions()
     });
   }
-
+  try{
+    // setting PWD back to working directory shouldn't be required
+    await fs.rename(tmpDirPath +'/'+name, workingDirPath+'/'+name)
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:66]: Please use `path.join` instead of concatenating strings. It would handle using the proper escaping and separators on all operating systems.

File: create-neon/src/shell.ts [hunk 5]
```diff
@@ -15,19 +15,38 @@ export default function shell(cmd: string, args: string[], cwd: string): Promise
     reject = rej;
   });
 
-  child.on('exit', (code) => {
+  function deleteNeonDir(): Promise<void> {
+    return fs.rm(cwd, { force: true, recursive: true });
```
Comment (kjvalencik) [inline create-neon/src/shell.ts:19]: Using `cwd` instead of a defined variable seems prone to error. What if we failed to changed directories?

File: create-neon/src/shell.ts [hunk 6]
```diff
@@ -15,19 +15,38 @@ export default function shell(cmd: string, args: string[], cwd: string): Promise
...
+      await deleteNeonDir();
       process.exit(code);
     }
+
+    if (code === 0) {
+      try {
+          let data = await fs.readFile(`${cwd}/package.json`,'utf8')
+          let { description, author, license } = JSON.parse(data);
+          if ([description, author, license].includes(undefined)) {
+               throw new Error("Ctrl+C pressed");
```
Comment (kjvalencik) [inline create-neon/src/shell.ts:37]: Is this error shown to the user? I think this could be multiple failures and not only `Ctrl+C`. Maybe something more generic?

```suggestion
               throw new Error("Failed to create node project with `npm init`");
```

File: create-neon/src/shell.ts [hunk 7]
```diff
@@ -15,19 +15,38 @@ export default function shell(cmd: string, args: string[], cwd: string): Promise
...
     if (code !== 0) {
+      await deleteNeonDir();
       process.exit(code);
     }
+
+    if (code === 0) {
+      try {
+          let data = await fs.readFile(`${cwd}/package.json`,'utf8')
+          let { description, author, license } = JSON.parse(data);
+          if ([description, author, license].includes(undefined)) {
```
Comment (kjvalencik) [inline create-neon/src/shell.ts:36]: Are these fields still defined if you hit `enter` without entering anything?

**Review by kjvalencik — CHANGES_REQUESTED** [review]

File: create-neon/src/bin/create-neon.ts [hunk 8]
```diff
@@ -30,7 +31,12 @@ function inferVersions(): Versions {
...
 
 async function main(name: string) {
+  let workingDirPath: string = process.cwd()
+  let tmpDirPath: string = tmpdir()
+
   try {
+    //change PWD to temp
+    process.chdir(tmpDirPath)
     await fs.mkdir(name);
   } catch (err) {
```
[the remark this answers is not in the document]
Comment (deepanshu44) [reply create-neon/src/bin/create-neon.ts:41]: A small question,
If file or folder exists, error message says- "file already exists"

Should we update the message to
"File or folder exists" 
Or
Should error message inform user that file, folder or symlink exists?

File: create-neon/src/shell.ts [hunk 9]
```diff
@@ -15,19 +15,38 @@ export default function shell(cmd: string, args: string[], cwd: string): Promise
...
 
-  child.on('exit', (code) => {
+  function deleteNeonDir(): Promise<void> {
+    return fs.rm(cwd, { force: true, recursive: true });
+  }
+
+  child.on("exit", async (code) => {
     if (code == null) {
-      process.exit(1);
+      await deleteNeonDir();
```
[the remark this answers is not in the document]
Comment (deepanshu44) [reply create-neon/src/shell.ts:24]: Alright, I will replace it with `die()`

File: create-neon/src/bin/create-neon.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kjvalencik) [reply create-neon/src/bin/create-neon.ts:41]: Current error looks like this, which is fine:

```
❯ npm init neon hello
❌ Could not create `hello`: EEXIST: file already exists, mkdir 'hello'
```

File: create-neon/src/shell.ts (hunk 7 above)
Comment (deepanshu44) [reply create-neon/src/shell.ts:36]: Yes, I tried doing the same with npm versions latest, LTS and a really old one 4.9.1 . Functionality is same and these fields do exist if we press `enter` without entering anything.

File: cli/src/async/child_process.ts [hunk 10]
```diff
@@ -2,11 +2,11 @@ import { promisify } from 'util';
 import child from 'child_process';
 
 export function spawn(command: string, args?: string[], options?: child.SpawnOptions): Promise<number> {
-  return new Promise((resolve, reject) => {
-      let ps = child.spawn(command, args || [], options!);
-      ps.on('error', reject);
-      ps.on('close', resolve);
-  });
+    return new Promise((resolve, reject) => {
```
Comment (kjvalencik) [inline cli/src/async/child_process.ts:5]: Please revert these changes. This project uses 2  space indentation.

File: create-neon/src/bin/create-neon.ts [hunk 11]
```diff
@@ -29,31 +30,60 @@ function inferVersions(): Versions {
   };
 }
 
+function deleteNeonDir(dir:string): Promise<void> {
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:33]: Nit, please leave a space after the `:` and before the type. This is repeated a few times in the PR.

Hopefully we can get `prettier` configured at some point!

File: create-neon/src/bin/create-neon.ts [hunk 12]
```diff
@@ -29,31 +30,60 @@ function inferVersions(): Versions {
...
+
 async function main(name: string) {
+  let workingDirPath: string = process.cwd()
+  let tmpDirPath: string = os.tmpdir()
+  let tmpFolderName:string|undefined;
+
   try {
-    await fs.mkdir(name);
+    //check if folder or file with same name exists
+    await fs.mkdir(name)
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:44]: Semicolons are missing in many places. Can you please add them? Again, sorry for the lack of automatic formatting!

File: create-neon/src/bin/create-neon.ts [hunk 13]
```diff
@@ -29,31 +30,60 @@ function inferVersions(): Versions {
...
+    await fs.rmdir(name)
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:45]: I like the `mkdir` and `rmdir` because it's a pretty lightweight way to check _both_ that it doesn't exist *and* that the user has write permissions.

Can you add a comment explaining this benefit?

File: create-neon/src/bin/create-neon.ts [hunk 14]
```diff
@@ -29,31 +30,60 @@ function inferVersions(): Versions {
...
+
+    //change current working directory to /temp
+    process.chdir(path.join(tmpDirPath))
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:48]: `path.join` is unnecessary here. However, my preference is to use an absolute path for this directory instead of relying on the current working directory.

Can you do something like this instead?

```js
let tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), `${name}-`));
```

File: create-neon/src/bin/create-neon.ts [hunk 15]
```diff
@@ -29,31 +30,60 @@ function inferVersions(): Versions {
...
+     await deleteNeonDir(path.join(tmpDirPath,tmpFolderName))
+    }
     die(`Could not create \`${name}\`: ${err.message}`);
   }
 
   let pkg: Package;
 
   try {
-    pkg = await Package.create(name);
+      pkg = await Package.create(name,path.join(tmpDirPath,tmpFolderName,'/'));
```
Comment (kjvalencik) [inline create-neon/src/bin/create-neon.ts:63]: Removing the trailing slash.

**Review by kjvalencik — CHANGES_REQUESTED** [review]

File: create-neon/src/shell.ts [hunk 16]
```diff
@@ -15,19 +16,31 @@ export default function shell(cmd: string, args: string[], cwd: string): Promise
...
-      process.exit(code);
-    }
+	reject(Error(`error code: ${code}`))
+    };
+
+    if (code === 0) {
+      try {
+          let data = await fs.readFile(path.join(cwd,'package.json'),'utf8')
+          let { description, author, license } = JSON.parse(data);
+          if ([description, author, license].includes(undefined)) {
```
Comment (kjvalencik) [inline create-neon/src/shell.ts:31]: This could benefit from a comment explaining that this is testing to ensure `npm init` was successful. It will catch `Ctrl+C`, answering `no` and many other failures.

Comment (deepanshu44) [reply create-neon/src/shell.ts:34]: See new [comments](https://github.com/neon-bindings/neon/commit/f7bb6e540301ce26638960dccf298e49650a9eb9)

**Comment by kjvalencik** [comment]
Thanks for all your hard work on this!
